akka-persistence-testkit/src/main/scala/akka/persistence/testkit/state/scaladsl/PersistenceTestKitDurableStateStore.scala

```diff
@@ -101,6 +107,62 @@ class PersistenceTestKitDurableStateStore[A](val system: ExtendedActorSystem)
       }, inclusive = true)
     }
 
+  override def currentChangesBySlices(
+      entityType: String,
+      minSlice: Int,
+      maxSlice: Int,
+      offset: Offset): Source[DurableStateChange[A], NotUsed] =
+    this.synchronized {
+      val currentGlobalOffset = lastGlobalOffset.get()
+      changesBySlices(entityType, minSlice, maxSlice, offset).takeWhile(_.offset match {
+        case Sequence(fromOffset) =>
+          fromOffset < currentGlobalOffset
+        case offset =>
+          throw new UnsupportedOperationException(s"$offset not supported in PersistenceTestKitDurableStateStore.")
+      }, inclusive = true)
+    }
+
+  override def changesBySlices(
+      entityType: String,
+      minSlice: Int,
+      maxSlice: Int,
+      offset: Offset): Source[DurableStateChange[A], NotUsed] =
+    this.synchronized {
+      val fromOffset = offset match {
+        case NoOffset             => EarliestOffset
+        case Sequence(fromOffset) => fromOffset
+        case offset =>
+          throw new UnsupportedOperationException(s"$offset not supported in PersistenceTestKitDurableStateStore.")
+      }
+      def bySliceFromOffset(rec: Record[A]) = {
+        val slice = persistence.sliceForPersistenceId(rec.persistenceId)
+        PersistenceId.extractEntityType(rec.persistenceId) == entityType && slice >= minSlice && slice <= maxSlice && rec.globalOffset > fromOffset
+      }
+      def bySliceFromOffsetNotDeleted(rec: Record[A]) =
+        bySliceFromOffset(rec) && store.contains(rec.persistenceId)
+
+      Source(store.values.toVector.filter(bySliceFromOffset).sortBy(_.globalOffset))
+        .concat(changesSource)
+        .filter(bySliceFromOffsetNotDeleted)
```

Comment:
This accesses mutable state from some other thread than the synchronized method call. Perhaps capturing the store as it was when the method was called is best? If not, make the `store` field volatile?

Reply:
@johanandren thanks, fixed in https://github.com/akka/akka/pull/30995/commits/b2bb9ef5b583cfbbf752c49dbd624aee9b03aa00